guest can add items to cart #11

julyytran opened the pull request:
*Created join table ShopItems to link shops and their items
*Created ShopController and Shop::ItemsController
*Created shop show that is effectively an item index for all items of that shop
*Created item show that is namespaced under its shop
*Structured it in this way so that a shop has its own specific url per the spec
*All tests passing

I should have made more commits; got too into it. 

Not 100% confident on the structuring of adding the shop layer in this way, but I arrived at this way through hopping down the error trail, and it works. Thoughts? @julsfelic @Unsafepond

julyytran committed:
guest can add items to cart

julsfelic commented on app/controllers/shop/items_controller.rb:
```diff
@@ -0,0 +1,5 @@
+class Shop::ItemsController < ApplicationController
+  def show
+    @item = find_item
```
Where does the `find_item` method live? What does it do?

julyytran replied:
It lives in the application controller, as a result of some refactoring we did towards the end

def find_item
    Item.find(params[:id])
 end

julsfelic commented on app/controllers/shops_controller.rb:
```diff
@@ -2,4 +2,9 @@ class ShopsController < ApplicationController
   end
+
+  def show
+    @shop = Shop.find(params[:id])
```
If you recall the one lesson we had where we wanted to find a shop by its slug? Let's say in theory we have a shop called Julian's Yummy Goods. We would end up wanting a URL like this: `/shops/julian-s-yummy-goods` There are a few steps to this, so I'll use specific code for each part. Within the `ShopsController` we want to be able to do `@shop = Shop.find_by(slug: params[:shop]`. In order to do this, we need to add a slug column to the `shops` table. Once we have that up, we can then create a callback method on the model `before_validation` (Don't remember if it is plural or not) that takes the name of the shop, uses the `pluralize` method to make a human readable URL and set that to the `slug` Below, I'll show you how the route would be setup

julsfelic commented on app/models/shop.rb:
```diff
@@ -1,2 +1,4 @@
 class Shop < ActiveRecord::Base
+  has_many :shop_items
+  has_many :items, through: :shop_items
```
In here is where we would set up the slug. Assuming we have the slug column, we could write something like this:

``` ruby
  before_validation :generate_slug

  private

  def generate_slug
    self.slug = name.parameterize
  end
```

julsfelic commented on app/views/shops/index.html.erb:
```diff
@@ -4,7 +4,7 @@
       <% @shops.each do |shop| %>
         <div id="shop">
-          <%= link_to shop.name, shops_path(shop) %>
+          <%= link_to shop.name, shop_path(shop) %>
```
I won't comment on each instance, but any view that is using `shop_path(shop)` would end up using `shop_path(shop: shop.slug)`

julsfelic commented on config/routes.rb:
```diff
@@ -14,7 +14,11 @@
-  resources :shops, only: [:index]
+  resources :shops, only: [:index, :show]
+
+  namespace :shop do
```
So to make everything above work we need to slightly alter this route. What we could do instead is:

``` ruby
  resources :shop, path: ":shop", as: :shop do
    ...
  end
```

julyytran replied:
Made all the slug changes, and am getting the error "No route matches {:action=>"show", :controller=>"shops", :shop=>"shop-1"} missing required keys: [:id]"

Example from class has the route
namespace :stores, path: ":store_name", as: :store do
....
end

but I also tried that and still get the same error. It is still expecting an id even though we want it to expect a slug, and where we change that expectation is in the route, yeah?

julyytran committed:
slug changes, routing error

julsfelic committed:
Add path for dynamic shop names

Since the shop names are dynamic and are start on the root, I had
to make a custom route to account for that.

julyytran committed:
correct plural shop_item controller to singular, drop shop items join table

julyytran commented:
Okay, finally, 
1. All tests passing!
2. Error was because it was looking for plural shops controller with item namespaced under it, but we had it as singular. Changed the route from plural to singular.
3. Changed the DB - dropped join table shop_items and put shop_id on items as primary key

Good to merge? @julsfelic @Unsafepond

julsfelic commented:
Everything is looking good to me! I'll go ahead and merge it.
